### PR TITLE
Added AbstractPickup, SpeedPickup and made them trigger when players collide with them.

### PR DESCRIPTION
**Pull request by joeyshuttleworth**

### CODE REVIEW

File: common/AbstractPickup.hpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#define PICKUP_SIZE 0.2
+
+#include "actor.hpp"
+#include "bloke.hpp"
+
+enum{
+     PICKUP_NONE,
+     PICKUP_SPEED,
+     PICKUP_BOMB
+};
```
Comment (tylerkf) [inline common/AbstractPickup.hpp:13]: 4-space indents when the rest of the document is 2-space, was this intentional?

File: common/GamePlayerProperties.hpp [hunk 2]
```diff
@@ -21,6 +21,14 @@ class GamePlayerProperties : public AbstractPlayerProperties{
     return;
   };
 
+  void reset(std::shared_ptr<AbstractPlayerProperties> p_properties){
+    std::shared_ptr<GamePlayerProperties> gp_properties = std::dynamic_pointer_cast<GamePlayerProperties>(p_properties);
+    if(gp_properties)
+      *this = GamePlayerProperties(*gp_properties);
+    else
+      *this = GamePlayerProperties();
+  }
```
Comment (tylerkf) [inline common/GamePlayerProperties.hpp:30]: Would be nice to add a brief comment describing the exact purpose of the function. i.e. to what extent does it "reset" the player and is it called at the end of each round or when you leave the game, etc.

File: common/bomb.cpp [hunk 3]
```diff
@@ -62,9 +62,11 @@ void bomb::update(){
 }
 
 void bomb::explode(){
+  remove();
   if(_server){
     /*Iterate over all actors and kill the ones if they are in the right (wrong) zone.*/
-      for(auto i = _pScene->mActors.begin(); i != _pScene->mActors.end(); i++){
+    const std::list<std::shared_ptr<actor>> actor_list = _pScene->mActors;
+      for(auto i = actor_list.begin(); i != actor_list.end(); i++){
```
Comment (tylerkf) [inline common/bomb.cpp:69]: Accidental indentation on for loop?

File: common/bomberbloke.h [hunk 4]
```diff
@@ -6,8 +6,10 @@
 #define DEFAULT_BLOKE_SIZE 0.9
 #define ACCELERATION_RATIO 0.5
 #define DECCELERATION_RATIO 0.1
-#define DEFAULT_MAX_SPEED double(0.5 / TICK_RATE)
+#define DEFAULT_SPEED double(0.5 / TICK_RATE)
+#define DEFAULT_MAX_SPEED double(0.1 / TICK_RATE)
```
Comment (tylerkf) [inline common/bomberbloke.h:10]: is max speed meant to be less than default speed?

File: common/woodenCrate.hpp [hunk 5]
```diff
@@ -19,10 +23,26 @@ class woodenCrate : public actor{
...
+        _pScene->addActor(std::make_shared<BombPickup>(mPosition[0], mPosition[1]));
+        break;
+      }
+      case PICKUP_BOMB:{
+        _pScene->addActor(std::make_shared<SpeedPickup>(mPosition[0], mPosition[1]));
+        break;
+      }
+      case PICKUP_NONE:
+      default: break;
+      }
```
Comment (tylerkf) [inline common/woodenCrate.hpp:45]: Would be nice if you left a brief comment, I don't think this mersenne twister thing makes it clear you're generating a random number. maybe it's standard code idk

File: engine/AbstractPlayerProperties.hpp [hunk 6]
```diff
@@ -4,7 +4,8 @@
 class AbstractPlayerProperties{
 public:
   //virtual void SetProperty() = 0;
-  virtual std::vector<std::string> GetProperties() = 0;
+  virtual std::vector<std::string> GetProperties(){return {};}
+  virtual void reset(std::shared_ptr<AbstractPlayerProperties> p_properties = nullptr) = 0;
```
Comment (tylerkf) [inline engine/AbstractPlayerProperties.hpp:8]: Comments?

File: engine/network/Interpolator.hpp [hunk 7]
```diff
@@ -84,7 +84,6 @@ class Interpolator{
     }
 
     if(state1 && !state2){
-      log_message(DEBUG, "No future state to interpolate, returning last known position");
```
Comment (tylerkf) [inline engine/network/Interpolator.hpp:87]: Accident? Just checking

File: engine/player/AbstractPlayer.cpp [hunk 8]
```diff
@@ -27,3 +27,14 @@ void AbstractPlayer::setId(int id){
...
+
+
+void AbstractPlayer::ResetPlayerProperties(std::shared_ptr<AbstractPlayerProperties> p_properties){
+   if(mpPlayerProperties){
+     /*  TODO: send EVENT_PROPERTIES message to the relevant player */
+     mpPlayerProperties->reset(p_properties);
+   }
+   else
+     log_message(ERROR, "Tried to reset mpPlayerProperties when it didn't exist");
+ }
```
Comment (tylerkf) [inline engine/player/AbstractPlayer.cpp:39]: Accidental 3-space indentation

File: engine/player/AbstractPlayer.hpp [hunk 9]
```diff
@@ -49,6 +50,9 @@ class AbstractPlayer{
 
   virtual int getPing(){return 0;}
 
+  void ResetPlayerProperties(std::shared_ptr<AbstractPlayerProperties> p_properties = nullptr);
```
Comment (tylerkf) [inline engine/player/AbstractPlayer.hpp:53]: Why switch out of camel-case?

**Review by tylerkf — APPROVED** [review]
Looks good! Check comments for some stylistic errors/suggestions

File: engine/network/Interpolator.hpp (hunk 7 above)
Comment (joeyshuttleworth) [reply engine/network/Interpolator.hpp:87]: There's not really any need for this line. It just spams the console too much.

File: engine/player/AbstractPlayer.hpp (hunk 9 above)
Comment (joeyshuttleworth) [reply engine/player/AbstractPlayer.hpp:53]: Amended

**Pull request merged**